From d1e70921625aaf3894de2f90d9aff75aa8b8a179 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 11 Jun 2024 13:57:03 +0200
Subject: [PATCH 0951/1695] drivers: stm32_risab: do not set CID configuration
 if not TDCID

Fix RISAB driver to set CID configuration only when OP-TEE is
TDCID since only TDCID can write to RISAB_PGy_CIDCFGR register. The CID
configuration must also be set before regs_access_granted() is called.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9ddd72f14210e4726fdebda7a14fa88dc43c9b6e drivers: stm32_risab: do not set CID configuration if not TDCID]
Change-Id: I1b4d7833aff5e755eb732d72d4bcad6a37303557
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/389862
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/firewall/stm32_risab.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/core/drivers/firewall/stm32_risab.c b/core/drivers/firewall/stm32_risab.c
index 51d4b4609..6f022a7fb 100644
--- a/core/drivers/firewall/stm32_risab.c
+++ b/core/drivers/firewall/stm32_risab.c
@@ -308,12 +308,16 @@ static void apply_rif_config(struct stm32_risab_pdata *risab_d)
 	for (i = 0; i < risab_d->nb_regions_cfged; i++) {
 		set_block_dprivcfgr(risab_d, i);
 
+		/* Delegate RIF configuration or not */
+		if (!is_tdcid)
+			DMSG("Cannot set %s CID configuration for region %u",
+			     risab_d->risab_name, i);
+		else
+			set_cidcfgr(risab_d, i);
+
 		if (!regs_access_granted(risab_d, i))
 			panic();
 
-		/* Delegate RIF configuration or not */
-		set_cidcfgr(risab_d, i);
-
 		/*
 		 * This sequence will generate an IAC if the CID filtering
 		 * configuration is inconsistent with these desired rights
@@ -550,11 +554,15 @@ static TEE_Result stm32_risab_pm_resume(struct stm32_risab_pdata *risab)
 		/* Restoring RISAB RIF configuration */
 		set_block_dprivcfgr(risab, i);
 
+		if (!is_tdcid)
+			DMSG("Cannot set %s CID configuration for region %zu",
+			     risab->risab_name, i);
+		else
+			set_cidcfgr(risab, i);
+
 		if (!regs_access_granted(risab, i))
 			continue;
 
-		set_cidcfgr(risab, i);
-
 		/*
 		 * This sequence will generate an IAC if the CID filtering
 		 * configuration is inconsistent with these desired rights
-- 
2.47.2

